Eliminate the use of oudated OpenSSL APIs

AndriiFullroot opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
Updating deprecated calls for OpenSSL

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
VD-2365

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
Those changes require full proper tests. I've tested authentication (chap, mschapv1, mschapv2).

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [ ] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

AndriiFullroot committed:
VD-2365: Update OpenSSL calls for MD4/MD5 and SHA1

Updating deprecated calls for hash calculation.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
VD-2365: Update OpenSSL DH usage

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
VD-2365: Update OpenSSL DES encryption

Updating deprecated calls for block encryption.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

ritika0313 commented on accel-pppd/auth/auth_mschap_v1.c:
```diff
@@ -418,15 +416,15 @@ static void des_encrypt(const uint8_t *input, const uint8_t *key, uint8_t *outpu
-	memcpy(cb, input, 8);
-	DES_ecb_encrypt(&cb, &res, &ks, DES_ENCRYPT);
-	memcpy(output, res, 8);
+	EVP_EncryptInit_ex(des_ctx, EVP_des_ecb(), NULL, cb, NULL);
```
By default, padding is enabled in EVP. Please consider :
1. Explicitly disabling the EVP padding (EVP_CIPHER_CTX_set_padding(ctx, 0)) 
2. It must be followed by a call to EVP_EncryptFinal_ex()

Currently things may work because the code encrypts exact 8-byte inputs and does not call EVP_EncryptFinal_ex, so padding is never added. But explicitly disabling padding avoids subtle bugs and makes code portable and future-proof

AndriiFullroot replied:
Yes, that's why I didn't call the `Final()` func. But may be its a good practice to "finish" the crypto.
`EVP_CIPHER_CTX_set_padding(ctx, 0)` - oh yeah, good point!.

ritika0313 commented on accel-pppd/extra/chap-secrets.c:
1. Line# 138:
Might want to remove EVP_MD_CTX_init(md_ctx) being an older API as per the information in below link:
https://docs.openssl.org/master/man7/ossl-guide-migration/#upgrading-from-openssl-102

2. Line# 717:
Please consider using EVP_MD_fetch() in place of EVP_get_digestbyname() as it is legacy API that returns a pointer to an EVP_MD; in OpenSSL 3 the provider model suggests using EVP_MD_fetch(libctx, name, propq) to allow provider-aware lookups. EVP_get_digestbyname() still works under compatibility/legacy provider, but is considered legacy for provider-based code.
For forward compatibility with OpenSSL 3's provider model, may replace calls with EVP_MD_fetch() and later do EVP_MD_free()

ritika0313 commented on accel-pppd/ctrl/l2tp/packet.c:
```diff
@@ -168,28 +168,31 @@ static int decode_avp(struct l2tp_avp_t *avp, const struct l2tp_attr_t *RV,
+	EVP_DigestFinal_ex(evp_ctx, p1, NULL);
 
-	if (attr_len <= MD5_DIGEST_LENGTH) {
+	if (attr_len <= EVP_MD_get_size(EVP_md5())) {
```
EVP_MD_get_size(EVP_md5()) is repeating, might want to calculate once and store in a variable

ritika0313 commented on accel-pppd/ctrl/l2tp/packet.c:
```diff
@@ -199,35 +202,37 @@ static int decode_avp(struct l2tp_avp_t *avp, const struct l2tp_attr_t *RV,
-	bytes_left = orig_attr_len + 2 - MD5_DIGEST_LENGTH;
-	last_block_len = bytes_left % MD5_DIGEST_LENGTH;
-	blocks_left = bytes_left / MD5_DIGEST_LENGTH;
+	bytes_left = orig_attr_len + 2 - EVP_MD_get_size(EVP_md5());
```
Same as previous:
EVP_MD_get_size(EVP_md5()) is repeating, might want to calculate once and store in a variable

AndriiFullroot replied:
Ok!

ritika0313 commented on accel-pppd/ctrl/sstp/sstp.c:
```diff
@@ -2584,21 +2584,20 @@ static void ssl_load_config(struct sstp_serv_t *serv, const char *servername)
 
-			SSL_CTX_set_tmp_dh(ssl_ctx, dh);
-			DH_free(dh);
+			SSL_CTX_set0_tmp_dh_pkey(ssl_ctx, dh);
```
Please consider checking the return value of SSL_CTX_set0_tmp_dh_pkey() to correctly to avoid memory leaks and surface configuration errors early.
```suggestion
	if (SSL_CTX_set0_tmp_dh_pkey(ssl_ctx, dhpkey) != 1) {
		log_error("sstp: %s error: %s\n",
		          "ssl-dhparam",
		          ERR_error_string(ERR_get_error(), NULL));
		EVP_PKEY_free(dhpkey);
		goto error;
	}
```

ritika0313 commented on accel-pppd/auth/auth_mschap_v1.c:
```diff
@@ -418,15 +416,15 @@ static void des_encrypt(const uint8_t *input, const uint8_t *key, uint8_t *outpu
-	DES_ecb_encrypt(&cb, &res, &ks, DES_ENCRYPT);
-	memcpy(output, res, 8);
+	EVP_EncryptInit_ex(des_ctx, EVP_des_ecb(), NULL, cb, NULL);
+	EVP_EncryptUpdate(des_ctx, output, NULL, input, EVP_CIPHER_block_size(EVP_des_ecb()));
```
Consider making use of output length parameter(outl). Passing it as NULL to EVP_EncryptUpdate(ctx, out, outl, in, inl) may be undefined on some builds as it expects a non-NULL value -_'The number of bytes written is placed in outl'_.

AndriiFullroot replied:
Ok!

ritika0313 commented on accel-pppd/auth/auth_mschap_v2.c:
```diff
@@ -500,18 +503,17 @@ static void des_encrypt(const uint8_t *input, const uint8_t *key, uint8_t *outpu
-	memcpy(cb,input,8);
-	DES_ecb_encrypt(&cb,&res,&ks,DES_ENCRYPT);
-	memcpy(output,res,8);
+	EVP_EncryptInit_ex(des_ctx, EVP_des_ecb(), NULL, cb, NULL);
```
Please consider to add padding disabling + outlen + final call (as mentioned previously)

AndriiFullroot replied:
Ok

ritika0313 commented on accel-pppd/extra/chap-secrets.c:
```diff
@@ -359,16 +357,15 @@ static void des_encrypt(const uint8_t *input, const uint8_t *key, uint8_t *outpu
-	memcpy(cb, input, 8);
-	DES_ecb_encrypt(&cb, &res, &ks, DES_ENCRYPT);
-	memcpy(output, res, 8);
+	EVP_EncryptInit_ex(evp_ctx, EVP_des_ecb(), NULL, cb, NULL);
```
Please consider to add padding disabling + outlen + final call (as mentioned previously)

AndriiFullroot replied:
Ok

ritika0313 reviewed (changes requested):
- I am not sure if the code completely relies on default/legacy provider. But in its abscence/failure to load it, EVP_des_ecb() will be NULL. Please evaluate if the scenario is applicable here or not. If it does, the code must handle the NULL case.
- For a safer code, please consider adding error handling and NULL checks wherever possible as EVP APIs have a possibility to fail.

AndriiFullroot commented on accel-pppd/auth/auth_mschap_v2.c:
```diff
@@ -500,18 +503,17 @@ static void des_encrypt(const uint8_t *input, const uint8_t *key, uint8_t *outpu
-	DES_ecb_encrypt(&cb,&res,&ks,DES_ENCRYPT);
-	memcpy(output,res,8);
+	EVP_EncryptInit_ex(des_ctx, EVP_des_ecb(), NULL, cb, NULL);
+	EVP_EncryptUpdate(des_ctx, output, NULL, input, EVP_CIPHER_block_size(EVP_des_ecb()));
```
Ok

AndriiFullroot commented on accel-pppd/auth/auth_mschap_v2.c:
```diff
@@ -608,32 +615,36 @@ static void set_mppe_keys(struct chap_auth_data *ad, uint8_t *z_hash, uint8_t *n
+	EVP_DigestUpdate(evp_ctx, pad1, sizeof(pad1));
+	EVP_DigestUpdate(evp_ctx, magic2, sizeof(magic2));
+	EVP_DigestUpdate(evp_ctx, pad2, sizeof(pad2));
+	EVP_DigestFinal_ex(evp_ctx, send_key, NULL);
```
Yes!

AndriiFullroot committed:
VD-2365: SSL calls refactoring

Added checks for context creation.
Added output length variables to En/DecryptUpdate calls.
Refactored code and added logs.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

ritika0313 commented on accel-pppd/extra/chap-secrets.c:
```diff
@@ -703,7 +748,7 @@ static void parse_hash_chain(const char *opt)
 		*ptr2 = 0;
 		hc = _malloc(sizeof(*hc));
-		hc->md = EVP_get_digestbyname(ptr1);
+		hc->md = EVP_MD_fetch(NULL, ptr1, NULL);
```
Please check, we might need to free the hc->md pointer using EVP_MD_free() in clear_hash_chain() and/or similar.

AndriiFullroot replied:
Ah, no EVP_MD doesn't require clearing, it's like calling EVP_sha1() etc.

dmbaturin reviewed (approved):
The change is straightforward, I believe we can merge it and give it some testing in VyOS rolling release images.
